src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs

```diff
@@ -105,7 +108,26 @@ public int Inflate(byte[] bytes, int offset, int length)
             int count = 0;
             do
             {
-                int copied = _output.CopyTo(bytes, offset, length);
+                int copied;
+                if (_uncompressedSize == -1)
+                {
+                    copied = _output.CopyTo(bytes, offset, length);
+                }
+                else
+                {
+                    if (_uncompressedSize > _currentInflatedCount)
+                    {
+                        copied = _output.CopyTo(bytes, offset, length);
```

Comment:
If I understand the flow correctly, the `copied` value is bounded by `length`.  Given that there's a `ReadByte()` method (which calls this with a length 1 already) the system can clearly handle `length` being "artificially low".

My concern with the current "just report a smaller number for `copied`" is that a call to Read which has a buffer bigger than the header's reported length can get back an answer like "1 byte was overwritten" when in reality the entire buffer was corrupted.

That's fine if `bytes` can never be a user-supplied buffer (that is, if it's only ever buffers that are held private by the implementation), but if it's a user-supplied buffer then we should limit the length value to `(int)(_uncompressedSize - _currentInflatedCount)`.  If that doesn't work, I'd really like to see what test is failing, because it suggests that there's a secondary problem in the implementation (because if this is user supplied buffers and lengths, the user could have already restricted the length themselves).

Reply:
Updated